python/ray/serve/_private/autoscaling_state.py

```diff
@@ -0,0 +1,355 @@
+        self._requests_queued_at_handles: Dict[str, float] = dict()
+        # Number of ongoing requests reported by replicas
+        self._replica_average_ongoing_requests: Dict[str, float] = dict()
+
+        self._deployment_info = None
+        self._autoscaling_policy_manager = None
+        self._running_replicas: List[ReplicaID] = []
+        self._target_capacity: Optional[float] = None
+        self._target_capacity_direction: Optional[TargetCapacityDirection] = None
+
+    @property
+    def autoscaling_policy_manager(self) -> AutoscalingPolicyManager:
+        return self._autoscaling_policy_manager
+
+    def register(
+        self, info: DeploymentInfo, curr_target_num_replicas: Optional[int] = None
+    ):
+        """Registers an autoscaling deployment's info."""
+
+        config = info.deployment_config.autoscaling_config
+        if (
+            self._deployment_info is None or self._deployment_info.config_changed(info)
+        ) and config.initial_replicas is not None:
+            target_num_replicas = config.initial_replicas
+        else:
+            target_num_replicas = curr_target_num_replicas
+
+        self._deployment_info = info
+        self._autoscaling_policy_manager = AutoscalingPolicyManager(config)
+        self._target_capacity = info.target_capacity
+        self._target_capacity_direction = info.target_capacity_direction
+
+        return self.apply_bounds(target_num_replicas)
+
+    def update_running_replica_ids(self, running_replicas: List[ReplicaID]):
+        """Update cached set of running replica IDs for this deployment."""
+        self._running_replicas = running_replicas
+
+    def is_within_bounds(self, num_replicas_running_at_target_version: int):
+        """Whether or not this deployment is within the autoscaling bounds.
+
+        Returns: True if the number of running replicas for the current
+            deployment version is within the autoscaling bounds. False
+            otherwise.
+        """
+
+        return (
+            self.apply_bounds(num_replicas_running_at_target_version)
+            == num_replicas_running_at_target_version
+        )
+
+    def apply_bounds(self, num_replicas: int):
+        """Clips a replica count with current autoscaling bounds.
+
+        This takes into account target capacity.
+        """
+
+        return self.autoscaling_policy_manager.apply_bounds(
+            num_replicas,
+            self._target_capacity,
+            self._target_capacity_direction,
+        )
+
+    def record_request_metrics_for_replica(
+        self, replica_id: ReplicaID, window_avg: Optional[float], send_timestamp: float
+    ) -> None:
+        """Records average number of ongoing requests at a replica."""
+
+        if window_avg is None:
+            return
+
+        if (
+            replica_id not in self._replica_average_ongoing_requests
+            or send_timestamp > self._replica_average_ongoing_requests[replica_id][0]
```

Comment:
> send_timestamp > self._replica_average_ongoing_requests[replica_id][0]

We should update [the type hint and comment](https://github.com/ray-project/ray/pull/44638/files#diff-47e03417ac344cec3fbcd930945c85aef89a773e22b59e314b6523c0f133a5dfR53) for `self._replica_average_ongoing_requests`. Those currently say that its values are the number of ongoing requests at a replica. We shouldn't be comparing a timestamp to that.